### PR TITLE
Front-End ED-to-FD Smart Contract Migration Tool 

### PR DESCRIPTION
**Pull request by Ebonsignori**

### CODE REVIEW

**Comment by activescott** [comment]
@Ebonsignori This is great. A couple comments and suggestions:

- Love that this is _not_ in main.js / main-unminified.js. Lets keep it that way. 👍 
- [ ] We should use the user's currently selected account and _not_ allow user to select account. You can get this from the [redux store](https://redux.js.org/api-reference/store) and this is a great way to decouple from main.js. 
    - To get to the redux store use `window.main.EtherDelta.store`
    - Calling [`store.getState`](https://redux.js.org/api-reference/store#getState) you'll see the state of the app and it is pretty easy to understand. For example, to get the account that the user currently has selected use `window.main.EtherDelta.store.getState().user.accounts[window.main.EtherDelta.store.getState().user.selectedAccount]` (obviously written verbosely for clarity).
- [ ] It would be great to keep the HTML portion of the UI out of index.html. Maybe we can put the UI in a separate JS file (react?), call into the JS and ask it to render into a given DOM element. Then in index.html we only have a placeholder element.
- [ ] Only ever show the migration popup if the user does have an ED balance.
- [ ] We should add a top-level menu item to access the utility
- We could potentially consider a separate repo for this if we componentize it properly. Then the new UI could use it too and it would force us to keep it cleanly designed. Curious what other son the @forkdelta/frontend-team  think about that.

Happy to work on this with you too. Let me know how I can best help.

**Comment by Ebonsignori** [comment]
@activescott, these are all excellent ideas.

Know that I've never worked with redux or react, but after the tutorials I've just reviewed, I agree that they are probably the best tools to use in this scenario (especially since the current front-end already uses them). 

With react, it should be simple to add a placeholder, `<div id="migrationTool"></div>` after the other placeholders in index.html, and render it from migrate.js. I'll attempt this, but could use a helpful code review to make sure that I followed the best practices. 

With redux, I see that code from main.js has already populated the store object with loads of useful information. **Does this information include orders and balances read from the ED smart contract?** If so, this makes the logic of migrate.js much simpler. If the store includes orders and balances, are they found as follows:
- In the toplevel `window.main.EtherDelta.store.getState()`
   - Are orders on the ED contract found in _myOrders_ ?
- In `window.main.EtherDelta.store.getState().user`
   - Are user balances on the ED contract  _user.accounts.balances_ ?

I'd greatly appreciate the help here, as I don't have an account with any order or balances  to test this.

Honestly, you are more knowledgeable of ForkDelta than I am, and your suggestions thus far should all be implemented, so any way that you want to contribute to / add to this project is up to you.

**Comment by JonathonDunford** [comment]
@Ebonsignori 

I wouldn't trust the UI to have the most up to date info. If you pull it from the utility contract, that's probably the most reliable source of information.

**Comment by JonathonDunford** [comment]
Notes moved over from Discord for posterity:

We should have a step 1, step 2
Step 1 is balances. Checks all tokens we list first, shows balances, asks if they have any tokens we didnt list, allows them to add new
keeps track of all balances

Then final part of step 1 is allowing users to select all to transfer everything (you can make this just console.log the balances if you don't really know how to sign the transactions and everything yet)

Step 2 would grab all orders from DB, list them in very similar way as step 1, allow users to replace all orders, pick and choose, or select none

Ideally all of the above would have progress sliders

== How to make transaction to sign for moving tokens ==


Basically, you will need to make 2 transactions for each token:
- Withdraw
- Deposit

You withdraw from old contract and deposit into new contract.

You can probably look at the code in UI to see how it's done, but basically to withdraw you invoke `withdraw` for ETH and `withdrawToken` for everything else for the old contract.

Then, for deposit, you invoke `deposit` for ETH and `depositToken` for everything else on the new contract.
UI will have the code for deposit/withdraw you can change up a bit.

**Comment by JonathonDunford** [comment]
For launch of this, we will need to:

- Move current orders database to orders_old database
- Clear out orders database
- Change logic on site to use new contract
- Deploy smart contract migration tool
- Add now "old" contract into dropdowns of contracts (remove the rest of them as they are unnecessary)
